### PR TITLE
[#11571] Selective Deadline Extensions - Add extension status to feedback session views

### LINKED ISSUE #11571: Selective Deadline Extensions
This issue keeps track of PRs relating to the selective deadline extensions feature. More info can be found on the wiki at https://github.com/TEAMMATES/teammates/wiki/Selective-Deadline-Extensions

We'll be working on a feature branch called [`selective-deadline-extension`](https://github.com/TEAMMATES/teammates/tree/selective-deadline-extension), and this issue will be resolved when the branch is merged into master. More details on maintaining a feature branch can be found in the [maintainer guide](https://github.com/TEAMMATES/teammates-ops/blob/master/maintainer-guide.md#using-a-long-lived-feature-branch).

### PR DESCRIPTION
**Pull request by FergusMok**
Part of #11571

With the new deadline fields in Feedback Sessions, each table row is able to parse through the maps to look for ongoing extensions, and display the appropriate status. The change appends "(Ext. ongoing)" to the status under "Submissions".

This change will affect two tables in 2 views, the Instructor's Home view, and the Instructor's Sessions view

https://user-images.githubusercontent.com/10182564/159123231-3c1927fb-5ac6-4b69-96b5-5b59199598f2.mp4

The longest possible string combination would be `Open (grace period) (Ext. ongoing)`

### CODE REVIEW

**Comment by moziliar** [comment]
A comment from @damithc in an earlier discussion: we can make the text "Closed (Ex. ongoing)" and use tool tip to explain further!

**Comment by FergusMok** [comment]
@moziliar Hi Zongran, I've changed the "with ongoing extensions" to "(Ext. ongoing)". The tooltip remains the same as in the video

File: src/web/app/components/teammates-common/submission-status-name.pipe.ts [hunk 1]
```diff
@@ -12,20 +12,49 @@ export class SubmissionStatusNamePipe implements PipeTransform {
...
+  }): boolean {
+    const timeNow = Date.now();
+
+    for (const studentEmail in deadlines.studentDeadlines) {
+      if (deadlines.studentDeadlines[studentEmail] > timeNow) return true;
+    }
+
+    for (const instructorDeadline in deadlines.instructorDeadlines) {
+      if (deadlines.studentDeadlines[instructorDeadline] > timeNow) return true;
+    }
```
Comment (moziliar) [inline src/web/app/components/teammates-common/submission-status-name.pipe.ts:56]: Can consider iterating with `Object.entries`

File: src/web/app/components/teammates-common/submission-status-name.pipe.ts [hunk 2]
```diff
@@ -12,20 +12,49 @@ export class SubmissionStatusNamePipe implements PipeTransform {
...
       case FeedbackSessionSubmissionStatus.CLOSED:
-        return 'Closed';
+        string += 'Closed';
+        break;
       default:
         return 'Unknown';
     }
-  }
 
+    if (deadlines && this.hasOngoingExtension(deadlines)) {
```
Comment (moziliar) [inline src/web/app/components/teammates-common/submission-status-name.pipe.ts:37]: Will Open be followed by Ex Ongoing?

File: src/web/app/components/sessions-table/submission-status-tooltip.pipe.ts [hunk 3]
```diff
@@ -47,9 +49,30 @@ export class SubmissionStatusTooltipPipe implements PipeTransform {
...
+    for (const studentEmail in deadlines.studentDeadlines) {
+      if (deadlines.studentDeadlines[studentEmail] > timeNow) return true;
+    }
+
+    for (const instructorDeadline in deadlines.instructorDeadlines) {
+      if (deadlines.studentDeadlines[instructorDeadline] > timeNow) return true;
+    }
+
+    return false;
+  }
```
Comment (moziliar) [inline src/web/app/components/sessions-table/submission-status-tooltip.pipe.ts:76]: Looks like a duplicated code in another file? Is there tear place to put it?

**Review by moziliar — CHANGES_REQUESTED** [review]
Looks great! Will try it out later!

File: src/web/app/components/sessions-table/submission-status-tooltip.pipe.ts (hunk 3 above)
Comment (FergusMok) [reply src/web/app/components/sessions-table/submission-status-tooltip.pipe.ts:76]: I'll refactor the pipe into the teammatescommon module, and put the helper method into a separate class

File: src/web/app/components/teammates-common/submission-status-name.pipe.ts (hunk 2 above)
Comment (FergusMok) [reply src/web/app/components/teammates-common/submission-status-name.pipe.ts:37]: Yup, I think it's ok to have "Open (Ext. ongoing)", as instructors may still want to see if there's extensions

Comment (moziliar) [reply src/web/app/components/teammates-common/submission-status-name.pipe.ts]: Sounds good

**Review by moziliar — CHANGES_REQUESTED** [review]
Generally LGTM. Good job! Can consider updating the video in the desc!

Fix lint too:)

**Review by moziliar — APPROVED** [review]
Generally looks good! I have nothing else to add.💯

**Comment by moziliar** [comment]
Rebase this too

**Comment by halfwhole** [comment]
Hey sorry could I trouble you to rebase once more? Seems to have some conflicts, esp. after #11712. Thanks!

File: src/web/services/deadline-extension-helper.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Deadline Extension utility functions.
+ */
+export class DeadlineExtensionHelper {
+  public static hasOngoingExtension(deadlines: {
+    studentDeadlines: Record<string, number>,
+    instructorDeadlines: Record<string, number>,
+  }): boolean {
+    const timeNow = Date.now();
+    const allDeadlines: Record<string, number> = { ...deadlines.studentDeadlines, ...deadlines.instructorDeadlines };
```
Comment (samuelfangjw) [inline src/web/services/deadline-extension-helper.ts:10]: Possible bug here. Suppose the following deadlines

```
studentDeadlines = {"abc@gmail.com": futureTime};
instructorDeadlines = {"abc@gmail.com": pastTime};
```

In this case, I believe the `abc@gmail.com`  in `instructorDeadlines` will override the one in `studentDeadlines` and give incorrect output.

Probably need to check both separately to be safe.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime());
+    const hasOngoingDeadlines = {
+      studentDeadlines: { ongoingDeadline: new Date('2021-01-01').valueOf() },
+      instructorDeadlines: { nonOngoingDeadline: new Date('2019-01-01').valueOf() },
+    };
+
+    const pipe: SubmissionStatusTooltipPipe = new SubmissionStatusTooltipPipe();
+    const extensionMessage = 'with current ongoing session extensions.';
+
+    const notVisibileWithExtension = pipe.transform(FeedbackSessionSubmissionStatus.NOT_VISIBLE, hasOngoingDeadlines);
```
Comment (samuelfangjw) [inline src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:41]: Minor typo in Visible, line 46 as well.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    expect(notVisibileWithExtension.substring(
+      notVisibileWithExtension.length - extensionMessage.length, notVisibileWithExtension.length),
+    ).toEqual(extensionMessage);
```
Comment (samuelfangjw) [inline src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:44]: Feeling fancy today :)

```suggestion
    const expectedExtensionMessage = expect.stringMatching(/.+current ongoing session extensions\.$/);
    expect(notVisibileWithExtension).toEqual(expectedExtensionMessage);
```

Alternatively, can assert `notVisibleWithExtension.endsWith(extensionMessage)` is true. I think both are easier to read than the current substring solution.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    };
+    const hasNoOngoingDeadlines = {
+      studentDeadlines: { nonOngoingExtension1: new Date('2019-01-01').valueOf() },
+      instructorDeadlines: { nonOngoingExtension2: new Date('2019-02-01').valueOf() },
+    };
+    const pipe: SubmissionStatusTooltipPipe = new SubmissionStatusTooltipPipe();
+
+    expect(pipe.transform(FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN, hasNoDeadlines)).toEqual(
+      pipe.transform(FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN),
+    );
```
Comment (samuelfangjw) [inline src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:24]: These should probably expect the output to be equal to a hardcoded string. The tests here should not just test that the behavior with and without extensions are equal, I don't think this test really makes sense without existing tests for the default behavior as well.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts [hunk 8]
```diff
@@ -1,8 +1,63 @@
...
+
+    const openWithExtension = pipe.transform(FeedbackSessionSubmissionStatus.OPEN, hasOngoingDeadlines);
+    expect(openWithExtension.substring(
+      openWithExtension.length - extensionMessage.length, openWithExtension.length),
+    ).toEqual(extensionMessage);
+
+    const gracePeriodWithExtension = pipe.transform(FeedbackSessionSubmissionStatus.GRACE_PERIOD, hasOngoingDeadlines);
+    expect(gracePeriodWithExtension.substring(
+      gracePeriodWithExtension.length - extensionMessage.length, gracePeriodWithExtension.length),
+    ).toEqual(extensionMessage);
```
Comment (moziliar) [inline src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts:56]: Will the use of `string.endswith` be more appropriate here? I see Samuel suggested something similar too.

File: src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts [hunk 9]
```diff
@@ -1,8 +1,63 @@
...
+
+    const closedWithExtension = pipe.transform(FeedbackSessionSubmissionStatus.CLOSED, hasOngoingDeadlines);
+    expect(closedWithExtension.substring(
```
Comment (moziliar) [inline src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts:59]: ditto

**Review by moziliar — CHANGES_REQUESTED** [review]
I have not much to add. The rest LGTM.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts (hunk 7 above)
Comment (FergusMok) [reply src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:24]: Makes sense, I've added another test for the default behavior too

Comment (FergusMok) [reply src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:24]: image

But testing for the default behavior through hardcoding does seem very redundant too. I'm thinking that the this test should be fine in serving it's purpose, but I'm open to suggestions

Comment (samuelfangjw) [reply src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:29]: Personally think it's fine.

File: src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts (hunk 8 above)
Comment (samuelfangjw) [reply src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts]: Agreed, actually using endsWith here might be better than the regex I suggested.

File: src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts [hunk 10]
```diff
@@ -1,8 +1,53 @@
...
+    jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime());
+    const hasNoDeadlines = {
+      studentDeadlines: {},
+      instructorDeadlines: {},
+    };
+    const hasNoOngoingDeadlines = {
+      studentDeadlines: { nonOngoingExtension: new Date('2019-01-01').valueOf() },
+      instructorDeadlines: {},
+    };
+    const pipe: SubmissionStatusNamePipe = new SubmissionStatusNamePipe();
```
Comment (jayasting98) [inline src/web/app/components/teammates-common/submission-status-name.pipe.spec.ts:20]: Perhaps initialise this in a `beforeEach`?

Probably the same for `SubmissionStatusTooltipPipe`.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+describe('SubmissionStatusTooltipPipe', () => {
+  it('create an instance', () => {
+    const pipe: SubmissionStatusTooltipPipe = new SubmissionStatusTooltipPipe();
+    expect(pipe).toBeTruthy();
+  });
+
+  it('transform with the tooltip description correctly', () => {
+    const pipe: SubmissionStatusTooltipPipe = new SubmissionStatusTooltipPipe();
+    expect(pipe).toBeTruthy();
+  });
```
Comment (jayasting98) [inline src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:13]: Did you forget to write this test case? It does not do what it says. Or is this superceded by `transform with deadlines correctly`? Maybe you forgot to remove?

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+  it('transform with deadlines correctly', () => {
+    jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime());
+    const hasOngoingDeadlines = {
+      studentDeadlines: { ongoingDeadline: new Date('2021-01-01').valueOf() },
+      instructorDeadlines: { nonOngoingDeadline: new Date('2019-01-01').valueOf() },
+    };
+
+    const pipe: SubmissionStatusTooltipPipe = new SubmissionStatusTooltipPipe();
+    const extensionMessage = 'with current ongoing individual deadline extensions.';
```
Comment (jayasting98) [inline src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:44]: Should there be a comma in front here too? Like here:
image

**Review by jayasting98 — COMMENTED** [review]
Sorry, just a few nits.

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts (hunk 11 above)
Comment (FergusMok) [reply src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:13]: Yup, sorry this was unintentional

File: src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts (hunk 12 above)
Comment (FergusMok) [reply src/web/app/components/teammates-common/submission-status-tooltip.pipe.spec.ts:44]: I think it won't matter, the test cases will correctly pass / fail without the whole string, but no harm to put for completion

**Review by jayasting98 — APPROVED** [review]
Thanks, I think it should be good to go already.

**Review by moziliar — APPROVED** [review]
LGTM!

**Pull request merged**